adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1320 @@
+		}
+	}
+	return adslot30, nil
+}
+
+func getNativeFormat(adslot30 *adslot30, openRTBImp *openrtb2.Imp) error {
+	if openRTBImp.Native.Request == "" {
+		return errors.New("extractAdmNative: imp.Native.Request is empty")
+	}
+
+	var nativePayload nativeRequests.Request
+	if err := json.Unmarshal(json.RawMessage(openRTBImp.Native.Request), &nativePayload); err != nil {
+		return err
+	}
+
+	var numImage = 0
+	var numVideo = 0
+	var width int64
+	var height int64
+	for _, asset := range nativePayload.Assets {
+		if asset.Video != nil {
+			numVideo++
+		}
+		// every image has the same W, H.
+		if asset.Img != nil {
+			numImage++
+			if asset.Img.H != 0 && asset.Img.W != 0 {
+				width = asset.Img.W
+				height = asset.Img.H
+			} else if asset.Img.WMin != 0 && asset.Img.HMin != 0 {
+				width = asset.Img.WMin
+				height = asset.Img.HMin
+			}
+		}
+	}
+	adslot30.W = width
+	adslot30.H = height
+
+	var detailedCreativeTypeList = make([]string, 0, 2)
+	if numVideo >= 1 {
+		detailedCreativeTypeList = append(detailedCreativeTypeList, "903")
+	} else if numImage > 1 {
+		detailedCreativeTypeList = append(detailedCreativeTypeList, "904")
+	} else if numImage == 1 {
+		detailedCreativeTypeList = append(detailedCreativeTypeList, "909")
+	} else {
+		detailedCreativeTypeList = append(detailedCreativeTypeList, "913", "914")
+	}
+	adslot30.DetailedCreativeTypeList = detailedCreativeTypeList
+	return nil
+}
+
+// convert adtype String -> Integer
+func convertAdtypeString2Integer(adtypeLower string) int32 {
+	if adtypeLower == "banner" {
+		return banner
+	} else if adtypeLower == "native" {
+		return native
+	} else if adtypeLower == "rewarded" {
+		return rewarded
+	} else if adtypeLower == "splash" {
+		return splash
+	} else if adtypeLower == "interstitial" {
+		return interstitial
+	} else if adtypeLower == "roll" {
+		return roll
+	} else {
+		return banner
+	}
+}
+
+// get app information for HuaweiAds request
+func getHuaweiAdsReqAppInfo(request *huaweiAdsRequest, openRTBRequest *openrtb2.BidRequest) error {
+	var app app
+	if openRTBRequest.App != nil {
+		if openRTBRequest.App.Ver != "" {
+			app.Version = openRTBRequest.App.Ver
+		}
+		if openRTBRequest.App.Name != "" {
+			app.Name = openRTBRequest.App.Name
+		}
+
+		// bundle cannot be empty, we need package name.
+		if openRTBRequest.App.Bundle != "" {
+			app.Pkgname = openRTBRequest.App.Bundle
+		} else {
+			return errors.New("HuaweiAdsReqApp: Pkgname is empty.")
+		}
+
+		if openRTBRequest.App.Content != nil && openRTBRequest.App.Content.Language != "" {
+			app.Lang = openRTBRequest.App.Content.Language
+		} else {
+			app.Lang = "en"
+		}
+	}
+	app.Country = getCountryCode(openRTBRequest)
+	request.App = app
+	return nil
+}
+
+// get field clientTime, format: 2006-01-02 15:04:05.000+0200
+func getClientTime(clientTime string) (newClientTime string) {
+	var zone = DefaultTimeZone
+	t := time.Now().Local().Format(time.RFC822Z)
```

Comment:
"ext": {
        "data": {
          "gaid": [
            "gaid"
          ],
          "oaid": [
            "oaid"
          ],
          "clientTime": [
            "2018-11-02 16:34:07.981+1300"
          ]
        }
      }
already tested in the test case，It’s best to use the time in this format from the media itself. If not, consider using the server time.

Reply:
Is the purpose of `t` now just to get a default timezone? If so then perhaps better to do it once on building your adapter, then save it in the struct and use adapter methods to access it rather than independent functions.

Reply:
In fact, when users use it, if they pass the time zone and the time under this time zone according to our requirements, our default time zone will not be used. This is for the user to be normal when the time zone is not passed.

Reply:
This default time zone will not be used under normal circumstances